File: examples/python/agg-inventory/handlers/state.py
"""Inventory state rebuilding logic."""

from dataclasses import dataclass, field

from google.protobuf.any_pb2 import Any as AnyProto

from angzarr import types_pb2 as types
from angzarr.state_builder import StateBuilder
from proto import inventory_pb2 as inventory


@dataclass
class InventoryState:
    product_id: str = ""
    on_hand: int = 0
    reserved: int = 0
    low_stock_threshold: int = 0
    reservations: dict = field(default_factory=dict)  # order_id -> quantity

    def exists(self) -> bool:
        return bool(self.product_id)

    def available(self) -> int:
        return self.on_hand - self.reserved

    def is_low_stock(self) -> bool:
        return self.available() < self.low_stock_threshold


# ============================================================================
# Named event appliers
# ============================================================================


def apply_stock_initialized(state: InventoryState, event: AnyProto) -> None:
    e = inventory.StockInitialized()
    event.Unpack(e)
    state.product_id = e.product_id
    state.on_hand = e.quantity
    state.reserved = 0
    state.low_stock_threshold = e.low_stock_threshold
    state.reservations = {}


def apply_stock_received(state: InventoryState, event: AnyProto) -> None:
    e = inventory.StockReceived()
    event.Unpack(e)
    state.on_hand = e.new_on_hand


def apply_stock_reserved(state: InventoryState, event: AnyProto) -> None:
    e = inventory.StockReserved()
    event.Unpack(e)
    state.on_hand = e.new_on_hand
    state.reserved = e.new_reserved
    state.reservations[e.order_id] = e.quantity


def apply_reservation_released(state: InventoryState, event: AnyProto) -> None:
    e = inventory.ReservationReleased()
    event.Unpack(e)
    state.on_hand = e.new_on_hand
    state.reserved = e.new_reserved
    state.reservations.pop(e.order_id, None)


def apply_reservation_committed(state: InventoryState, event: AnyProto) -> None:
    e = inventory.ReservationCommitted()
    event.Unpack(e)
    state.on_hand = e.new_on_hand
    state.reserved = e.new_reserved
    state.reservations.pop(e.order_id, None)


# ============================================================================
# State rebuilding
# ============================================================================

# stateBuilder is the single source of truth for event type -> applier mapping.
_state_builder = (
    StateBuilder(InventoryState)
    .on("StockInitialized", apply_stock_initialized)
    .on("StockReceived", apply_stock_received)
    .on("StockReserved", apply_stock_reserved)
    .on("ReservationReleased", apply_reservation_released)
    .on("ReservationCommitted", apply_reservation_committed)
)


def rebuild_state(event_book: types.EventBook | None) -> InventoryState:
    """Rebuild inventory state from events."""
    return _state_builder.rebuild(event_book)


def apply_event(state: InventoryState, event: AnyProto) -> None:
    """Apply a single event to the inventory state."""
    _state_builder.apply(state, event)

File: examples/go/agg-inventory/logic/state.go
package logic

import (
	"angzarr"
	angzarrpb "angzarr/proto/angzarr"
	"angzarr/proto/examples"

	"google.golang.org/protobuf/types/known/anypb"
)

// ============================================================================
// Named event appliers
// ============================================================================

func applyStockInitialized(state *InventoryState, event *anypb.Any) {
	var e examples.StockInitialized
	if err := event.UnmarshalTo(&e); err != nil {
		return
	}
	state.ProductID = e.ProductId
	state.OnHand = e.Quantity
	state.Reserved = 0
	state.LowStockThreshold = e.LowStockThreshold
	state.Reservations = make(map[string]int32)
}

func applyStockReceived(state *InventoryState, event *anypb.Any) {
	var e examples.StockReceived
	if err := event.UnmarshalTo(&e); err != nil {
		return
	}
	state.OnHand = e.NewOnHand
}

func applyStockReserved(state *InventoryState, event *anypb.Any) {
	var e examples.StockReserved
	if err := event.UnmarshalTo(&e); err != nil {
		return
	}
	state.OnHand = e.NewOnHand
	state.Reserved = e.NewReserved
	if state.Reservations == nil {
		state.Reservations = make(map[string]int32)
	}
	state.Reservations[e.OrderId] = e.Quantity
}

func applyReservationReleased(state *InventoryState, event *anypb.Any) {
	var e examples.ReservationReleased
	if err := event.UnmarshalTo(&e); err != nil {
		return
	}
	state.OnHand = e.NewOnHand
	state.Reserved = e.NewReserved
	delete(state.Reservations, e.OrderId)
}

func applyReservationCommitted(state *InventoryState, event *anypb.Any) {
	var e examples.ReservationCommitted
	if err := event.UnmarshalTo(&e); err != nil {
		return
	}
	state.OnHand = e.NewOnHand
	state.Reserved = e.NewReserved
	delete(state.Reservations, e.OrderId)
}

func loadInventorySnapshot(state *InventoryState, snapshot *anypb.Any) {
	var snapState examples.InventoryState
	if err := snapshot.UnmarshalTo(&snapState); err != nil {
		return
	}
	state.ProductID = snapState.ProductId
	state.OnHand = snapState.OnHand
	state.Reserved = snapState.Reserved
	state.LowStockThreshold = snapState.LowStockThreshold
	state.Reservations = make(map[string]int32)
	for k, v := range snapState.Reservations {
		state.Reservations[k] = v
	}
}

// ============================================================================
// State rebuilding
// ============================================================================

// stateBuilder is the single source of truth for event type → applier mapping.
var stateBuilder = angzarr.NewStateBuilder(func() InventoryState {
	return InventoryState{Reservations: make(map[string]int32)}
}).
	WithSnapshot(loadInventorySnapshot).
	On("StockInitialized", applyStockInitialized).
	On("StockReceived", applyStockReceived).
	On("StockReserved", applyStockReserved).
	On("ReservationReleased", applyReservationReleased).
	On("ReservationCommitted", applyReservationCommitted)

// RebuildState reconstructs inventory state from an event book.
func RebuildState(eventBook *angzarrpb.EventBook) InventoryState {
	return stateBuilder.Rebuild(eventBook)
}

// ApplyEvent applies a single event to the inventory state.
func ApplyEvent(state *InventoryState, event *anypb.Any) {
	stateBuilder.Apply(state, event)
}
